lib/build-utilities/migrate-initializers.js

```diff
@@ -0,0 +1,127 @@
+                         `Please refer to www.ember-fastboot.com for a migration path.`;
+    throw new Error(errorMsg);
+  }
+}
+
+/**
+ * Function to move the fastboot initializers to fastboot/app
+ * @param {Object} project
+ * @param {String} rootPath
+ */
+function _moveFastBootInitializers(project, rootPath) {
+  const deprecate = project.ui.writeDeprecateLine.bind(project.ui);
+
+  // check to see if it is a fastboot complaint addon
+  const isFastbootAddon = _checkInitializerTypeExists(rootPath, 'fastboot');
+  if (isFastbootAddon) {
+    deprecate(`Having fastboot specific code in app directory of ${rootPath} is deprecated. Please move it to fastboot/app directory.`);
+
+    const fastbootDirPath = path.join(rootPath, FASTBOOT_DIR);
+    // check if fastboot/app exists
+    if(!existsSync(fastbootDirPath)) {
+      fs.mkdirsSync(fastbootDirPath);
+    }
+
+    // copy over app/initializers/fastboot and app/instance/initializers/fastboot
+    fastbootInitializerTypes.forEach((fastbootInitializerType) => {
+      var srcFastbootPath = path.join(rootPath, 'app', fastbootInitializerType, 'fastboot');
+
+      if (existsSync(srcFastbootPath)) {
+        var destFastbootPath = path.join(fastbootDirPath, fastbootInitializerType);
+        if (!existsSync(destFastbootPath)) {
+          fs.mkdirSync(destFastbootPath);
+        }
+
+        // fastboot initializer type exists so we need to move this fastboot/app
+        var fastbootFiles = fs.readdirSync(srcFastbootPath);
+        fastbootFiles.forEach((fastbootFile) => {
+          var srcPath = path.join(srcFastbootPath, fastbootFile);
+          var destPath = path.join(destFastbootPath, fastbootFile);
+          fs.copySync(srcPath, destPath);
+
+          // delete the original path files so that there are no two initializers with the same name
+          fs.unlinkSync(srcPath);
+        });
+      }
+    });
+  }
+}
+
+/**
+ * Function that migrates the fastboot initializers for all addons.
+ *
+ * @param {Object} project
+ */
+function _migrateAddonInitializers(project) {
+  const nodeModulesPath = path.join(project.nodeModulesPath);
+
+  project.addons.forEach((addon) => {
+    const currentAddonPath = path.join(nodeModulesPath, addon.name);
+
+    _checkBrowserInitializers(currentAddonPath);
+    _moveFastBootInitializers(project, currentAddonPath);
+  });
+}
+
+/**
+ * Function to migrate fastboot initializers for host app.
+ *
+ * @param {Object} project
+ */
+function _migrateHostAppInitializers(project) {
+  const hostAppPath = path.join(project.root);
+
+  _checkBrowserInitializers(hostAppPath);
+  _moveFastBootInitializers(project, hostAppPath);
+}
+
+/**
+ * Function that migrates all addons and host app fastboot initializers to fastboot/app.
+ * It also throws an error if any addon or host app has browser forked initializers.
+ *
+ * @param {Object} project
+ */
+function migrateInitializers(project) {
```

Comment:
@simonihmig wouldn't moving in broccoli-tree be very expensive? Are you talking about after the `treeForApp` is generated? I don't think we will solve the build performance problem then (at least in a temporary sense). 

I totally agree it is not the best way and I am happy to explore new ways to do it. Do you have an example that you can explain with?

Reason for changing the original in-repo files is (maybe moving in broccoli tree solves the below issue):
the problem not moving in-repo files is (as I mentioned earlier) that if we don't move you will end up with two AMD modules after the build:
1. `<app-name>/initializers/fastboot/x.js` <-- This one will get loaded in browser and FastBoot
2. `<app-name>/initializers/x.js` <--- this one is only loaded in FastBoot

When the app will boot in the browser, it _will_ run `app/initializers/fastboot/x.js` (this because we no longer filter the fastboot initializers) which will cause issues. Hence there is a need to _move_ the in-repo files so that we don't generate `<app-name>/initializers/fastboot/x.js`. 

Maybe there is a better way to solve this problem that I am not seeing?